ros/angel_system_nodes/angel_system_nodes/hmm.py

```diff
@@ -0,0 +1,254 @@
+
+            # Add activity classification to the HMM
+            self._hmm.add_activity_classification(
+                activity_msg.label_vec,
+                activity_msg.conf_vec,
+                source_stamp_start_frame_sec,
+                source_stamp_end_frame_sec,
+            )
+
+            # Tell the HMM threads to wake up
+            self._hmm_state_awake_evt.set()
+            self._hmm_skip_awake_evt.set()
+
+    def publish_task_state_message(self, activity=None):
+        """
+        Forms and sends a `angel_msgs/TaskUpdate` message to the
+        TaskUpdates topic.
+        """
+        log = self.get_logger()
+
+        message = TaskUpdate()
+
+        # Populate message header
+        message.header.stamp = self.get_clock().now().to_msg()
+        message.header.frame_id = "Task message"
+
+        # Populate steps and current step
+        message.steps = self._hmm.model.class_str
+        message.current_step = self._current_step
+        message.current_step_id = self._hmm.model.class_str.index(self._current_step)
+
+        if self._previous_step is None:
+            message.previous_step = "N/A"
+        else:
+            message.previous_step = self._previous_step
+
+        # TODO: Populate task name and description
+        # TODO: Do we need to fill in current/next activity?
+        # TODO: Fill in time remaining?
+
+        self._publisher.publish(message)
+
+        # Update state tracking vars
+        self._previous_step = self._current_step
+
+    def query_task_graph_callback(self, request, response):
+        """
+        Populate the `QueryTaskGraph` response with the task list
+        and return it.
+        """
+        log = self.get_logger()
+        task_g = TaskGraph()
+
+        task_g.task_steps = self._hmm.model.class_str
+
+        # TODO: support different task levels?
+        task_g.task_levels = [0] * len(self._hmm.model.class_str)
+
+        response.task_graph = task_g
+        return response
+
+    def thread_run_hmm_state(self):
+        """
+        HMM runtime thread that gets the current state.
+        Thread is awakened each time a new ActivityDetection message is
+        received.
+        """
+        log = self.get_logger()
+        log.info("HMM state thread started")
+
+        while self._hmm_active.wait(0): # will quickly return false if cleared
+            if self._hmm_state_awake_evt.wait(self._hmm_active_heartbeat):
+                log.info("HMM state loop awakened")
+                self._hmm_state_awake_evt.clear()
+
+                # Get the HMM prediction
+                start_time = time.time()
+                self._current_step = self._hmm.get_current_state()
+                log.info(f"HMM get_current_state time: {time.time() - start_time}")
+                log.info(f"Current step {self._current_step}")
+
+                # Publish a new TaskUpdate message
+                self.publish_task_state_message()
+
+    def thread_run_hmm_skip(self):
```

Comment:
Does this need to be a separate thread? If so, the rationale should be included in the description of the method.